sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSortExec.scala

```diff
@@ -16,254 +16,459 @@
+    val outputRows = gpuLongMetric(NUM_OUTPUT_ROWS)
+    val outOfCore = sortType == OutOfCoreSort
+    child.executeColumnar().mapPartitions { cbIter =>
+      if (outOfCore) {
+        val cpuOrd = new LazilyGeneratedOrdering(sorter.cpuOrdering)
+        val spillCallback = GpuMetric.makeSpillCallback(allMetrics)
+        val iter = GpuOutOfCoreSortIterator(cbIter, sorter, cpuOrd,
+          // bad things can happen if the target size is really small
+          math.max(16 * 1024, targetSize), totalTime, sortTime, outputBatch, outputRows,
+          peakDevMemory, spillCallback)
+        TaskContext.get().addTaskCompletionListener(_ -> iter.close())
+        iter
+      } else {
+        GpuSortEachBatchIterator(cbIter, sorter, totalTime, sortTime, outputBatch, outputRows,
+          peakDevMemory)
+      }
     }
   }
+}
 
-  private def createBatchGpuSorter(): GpuColumnarBatchSorter = {
-    val boundSortExprs = GpuBindReferences.bindReferences(sortOrder, output)
-    new GpuColumnarBatchSorter(boundSortExprs, this, coalesceGoal == RequireSingleBatch)
+case class GpuSortEachBatchIterator(
+    iter: Iterator[ColumnarBatch],
+    sorter: GpuSorter,
+    totalTime: GpuMetric = NoopMetric,
+    sortTime: GpuMetric = NoopMetric,
+    outputBatches: GpuMetric = NoopMetric,
+    outputRows: GpuMetric = NoopMetric,
+    peakDevMemory: GpuMetric = NoopMetric) extends Iterator[ColumnarBatch] with Arm {
+  override def hasNext: Boolean = iter.hasNext
+
+  override def next(): ColumnarBatch = {
+    withResource(iter.next()) { cb =>
+        withResource(new NvtxWithMetrics("sort total", NvtxColor.WHITE, totalTime)) { _ =>
+          val ret = sorter.fullySortBatch(cb, sortTime, peakDevMemory)
+          outputBatches += 1
+          outputRows += ret.numRows()
+          ret
+        }
+    }
   }
 }
 
-class GpuColumnarBatchSorter(
-    sortOrder: Seq[SortOrder],
-    exec: GpuExec,
-    singleBatchOnly: Boolean,
-    shouldUpdateMetrics: Boolean = true) extends Serializable {
-
-  private var totalSortTimeNanos = 0L
-  private var maxDeviceMemory = 0L
-  private var haveSortedBatch = false
-  private val numSortCols = sortOrder.length
-  private val totalTimeMetric : Option[GpuMetric] = initMetric(TOTAL_TIME)
-  private val outputBatchesMetric : Option[GpuMetric] = initMetric(NUM_OUTPUT_BATCHES)
-  private val outputRowsMetric : Option[GpuMetric] = initMetric(NUM_OUTPUT_ROWS)
-
-  private def initMetric(metricName: String): Option[GpuMetric] = if (shouldUpdateMetrics) {
-    Some(exec.gpuLongMetric(metricName))
-  } else {
-    None
+case class OutOfCoreBatch(buffer: SpillableColumnarBatch, row: UnsafeRow) extends AutoCloseable {
+  override def close(): Unit = buffer.close()
+}
+
+class Pending(cpuOrd: LazilyGeneratedOrdering) extends AutoCloseable {
+  private val pending = new PriorityQueue[OutOfCoreBatch](new Comparator[OutOfCoreBatch]() {
+    override def compare(a: OutOfCoreBatch, b: OutOfCoreBatch): Int =
+      cpuOrd.compare(a.row, b.row)
```

Comment:
Could `row` be renamed to something like `cutOff` or `lower`, if I understand this correctly?